evm/src/common.rs

```diff
@@ -0,0 +1,116 @@
+extern crate hex;
+
+use ethereum_types::*;
+use std::u64;
+
+#[inline]
+pub fn to_word_size(size: u64) -> u64 {
+    if size > u64::MAX - 31 {
+        return u64::MAX >> 5 + 1;
+    }
+    return (size + 31) >> 5;
+}
+
+#[inline]
+pub fn mem_gas_cost(size: u64, memory_gas: u64) -> u64 {
+    let size = to_word_size(size);
+    size * memory_gas + ((size * size) >> 9)
+}
+
+#[inline]
+pub fn get_sign(value: U256) -> (U256, bool) {
+    let U256(arr) = value;
+    let sign = arr[3].leading_zeros() == 0;
```

Comment:
这里是什么意思，能帮忙介绍下么？

Reply:
U256 是无符号整数, 这里将 U256 解析为有符号整数. 类似 u64 -> i64 这样的解析

Reply:
范围变为 -2^255 ~ 2^255-1

Reply:
`arr[3].leading_zeros()` 这个是怎么算出来的？

Reply:
em。明白了。看了下这个类型提供的leading_zero，https://github.com/paritytech/parity-common/blob/048951f0c32d193bdc402e8e67ccb55d5c539ebd/uint/src/uint.rs#L486

应该是可以直接这样用的，用这种“黑魔法”需要知道底层的实现。换成这种方式是否可以？

Reply:
对性能会有影响. 比如一个 U256 前 255 位都是 0, 那直接 U256.leading_zeros 需要完全读取 4 个 u64 的值并返回 255, 但事实上只有最后一个 u64 是我们需要的.

Reply:
不过要不要扣这种细节真不好说... 代码可读性与性能之间得有家伙要妥协...

Reply:
你觉得如何? 改还是不改

Reply:
加一个注释吧，可以不改，这个应该是底层类型支持的。

Reply:
@yangby-cryptape

Reply:
For readability only, low performance for this U256 implementation: `value >> 255 == 0`.

Reply:
`arr[3] >> 63`

Reply:
好的, 我修改一下

Reply:
而且这个 leading_zeros 意义挺不明确的。。

Reply:
arr[3] >> 63 == 1

Reply:
[leading_zeros](https://doc.rust-lang.org/std/primitive.u64.html#method.leading_zeros) is a built-in function for rust primitive integer type.
I have implemented it for my fixed uints, too.

I prefer to `value >> 255 == 1` rather than `arr[3] >> 63 == 1`.
In some version, the `U256` here used was `U256([u32; 8])`, but now, it is `U256([u64; 4])`, may be one day, it was changed to `U256([u128; 2])`.
This piece of codes was depended on the specific implementation.

Reply:
Yes, we have talked `leading_zeros ` before. See above https://github.com/paritytech/parity-common/blob/048951f0c32d193bdc402e8e67ccb55d5c539ebd/uint/src/uint.rs#L486.
And I think the `leading_zeros` is not explicit.
I am not sure whether `value >> 255` is correct.

Reply:
https://github.com/paritytech/parity-common/blob/048951f0c32d193bdc402e8e67ccb55d5c539ebd/uint/src/uint.rs#L1150

We both agreed to `value >> 255` now.